### PR TITLE
Request validation plugin

### LINKED ISSUE #1643: request help: How to validate request in APISIX
### Issue description
I would like to validate requests before sending to up-streams, for example:

```
header_data = {
  auth_header = {required, type = 'string'}
}
body_data = {
  payload_1 = {required, type = 'boolean'},
  payload_2 = {optional}
}
```

### PR DESCRIPTION
**Pull request by sshniro**
Resolves #1643 
The plugin uses the `json-schema` validator to validate requests before sending them to upstream.
This plugin can be used to validate the header and body data.

### CODE REVIEW

**Comment by sshniro** [comment]
@membphis , @moonming should I write more test cases for different types of schema validations or rely on the JSON schema validator itself?

File: apisix/plugins/request-validation.lua [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            core.log.error("req schema validation failed", err)
+            core.response.exit(400, err)
+        end
+    end
+
+    if not conf.body_schema.properties.body_schema then
+        ngx.req.read_body()
+        local body = ngx.req.get_body_data()
+        if not body then
+            body = ngx.req.get_body_file()
```
Comment (membphis) [inline apisix/plugins/request-validation.lua:64]: the current way is wrong.

`file_name = ngx.req.get_body_file()`

https://github.com/openresty/lua-nginx-module#ngxreqget_body_file

File: apisix/plugins/request-validation.lua [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        end
+
+        local data, error = json_decode(body)
```
Comment (membphis) [inline apisix/plugins/request-validation.lua:67]: we should check the `Content-Type` header first.

https://stackoverflow.com/questions/20620300/http-content-type-header-and-json

File: apisix/plugins/request-validation.lua [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+local ngx           = ngx
+
+
+local schema = {
+    type = "object",
+    properties = {
+        body_schema = {type = "object"},
+        header_schema = {type = "object"}
+    },
+    oneOf = {
```
Comment (membphis) [inline apisix/plugins/request-validation.lua:29]: I think we should use `anyOf`, we can enable them at the same time

File: apisix/plugins/request-validation.lua (hunk 1 above)
Comment (sshniro) [reply apisix/plugins/request-validation.lua:64]: Shall I not read the body file, as its anyways going to be big? The following is the sample code I got to read the body file. And it seems, reading file will block the even loop of nginx.
```lua
local filename = ngx.req.get_body_file()
if not filename then
    return ngx.exit(ngx.HTTP_INTERNAL_SERVER_ERROR)
end
-- WARNING
-- don't use this code in production, file I/O is blocking,
-- you are going to block nginx event loop at this point!
local fd = io.open(filename, 'rb')
if not fd then
    return ngx.exit(ngx.HTTP_INTERNAL_SERVER_ERROR)
end
body = fd:read('*a')
```

File: apisix/plugins/request-validation.lua (hunk 2 above)
Comment (sshniro) [reply apisix/plugins/request-validation.lua:67]: @membphis what would be the behavior if the header is not found, reject the call with `400`?

File: apisix/plugins/request-validation.lua (hunk 1 above)
Comment (membphis) [reply apisix/plugins/request-validation.lua]: if we use `jsonschema` to check the request body, we have to do it in this way.

File: apisix/plugins/request-validation.lua (hunk 2 above)
Comment (membphis) [reply apisix/plugins/request-validation.lua]: https://stackoverflow.com/questions/9870523/differences-in-application-json-and-application-x-www-form-urlencoded

I think we can use default value: `application/json` if the header is not found.

if the header `Content-Type` is `application/x-www-form-urlencoded`, then we should use https://github.com/openresty/lua-nginx-module#ngxdecode_args .

File: apisix/plugins/request-validation.lua [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                  core.response.exit(400)
+                  return
+                end
+
+                local ok, err = core.schema.check(conf.body_schema, data)
+                if not ok then
+                  core.log.error("req schema validation failed", err)
+                  core.response.exit(400, err)
+                end
+              end
```
Comment (membphis) [inline apisix/plugins/request-validation.lua:79]: bad indentation

File: apisix/plugins/request-validation.lua [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            core.log.error("req schema validation failed", err)
+            core.response.exit(400, err)
+        end
+    end
+
+    if not conf.body_schema.properties.body_schema then
+        ngx.req.read_body()
+        local body = ngx.req.get_body_data()
+
+        if headers["content-type"] then
```
Comment (membphis) [inline apisix/plugins/request-validation.lua:64]: how about this style?

```
ngx.req.read_body()
local body = ngx.req.get_body_data()
local req_body, err

if headers["content-type"] == "application/x-www-form-urlencoded" then
    req_body, err = ngx.decode_args(body)
else -- JSON as default
    req_body, err = core.json.decode(body)
end

if not req_body then
    ... ...
    return 
end

local ok, err = core.schema.check(conf.body_schema, req_body)
...
```

File: apisix/plugins/request-validation.lua [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    version = 0.1,
+    priority = 2800,
+    type = 'validation',
+    name = plugin_name,
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
```
Comment (membphis) [inline apisix/plugins/request-validation.lua:45]: add a todo comment: `need a way to check if the input scheme is a valid Jsonschema`

File: doc/plugins/request-validation.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+## How To Enable
+
+Create a route and enable the request-validation plugin on the route:
+
+```shell
+curl http://127.0.0.1:9080/apisix/admin/routes/5 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+{
+    "uri": "/get",
+    "plugins": {
+        "body_schema": {
```
Comment (membphis) [inline doc/plugins/request-validation.md:54]: the plugin name is wrong

File: apisix/plugins/request-validation.lua [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+end
+
+
+function _M.rewrite(conf)
+    local headers = ngx.req.get_headers()
+
+    if conf.body_schema.properties.header_schema then
```
Comment (membphis) [inline apisix/plugins/request-validation.lua:52]: is it should be `if conf.header_schema then` ?

File: apisix/plugins/request-validation.lua [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        local ok, err = core.schema.check(conf.header_schema, headers)
+        if not ok then
+            core.log.error("req schema validation failed", err)
+            core.response.exit(400, err)
+        end
+    end
+
+    if not conf.body_schema.properties.body_schema then
```
Comment (membphis) [inline apisix/plugins/request-validation.lua:60]: is it should be `if conf. body_schema then` ?

File: apisix/plugins/request-validation.lua [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        {required = {"body_schema"}},
+        {required = {"header_schema"}}
+    }
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = 2800,
+    type = 'validation',
```
Comment (membphis) [inline apisix/plugins/request-validation.lua:38]: the default type is enough.

File: apisix/plugins/request-validation.lua (hunk 1 above)
Comment (shenal) [reply apisix/plugins/request-validation.lua]: If would be nice if a json formatted error is returned, Rather than the generic page

**Comment by sshniro** [comment]
Hi All, I will address the review comments for this plugin in a week ( laptop is repair).

File: apisix/plugins/request-validation.lua (hunk 8 above)
Comment (sshniro) [reply apisix/plugins/request-validation.lua]: correct, fixed.

File: apisix/plugins/request-validation.lua (hunk 10 above)
Comment (sshniro) [reply apisix/plugins/request-validation.lua:38]: removed.

File: t/plugin/request-validation.t [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+        }
+    }
+--- request
+GET /t
+--- response_body
+required field missing
+--- error_log
+property "required_payload" is required
+
+
```
Comment (membphis) [inline t/plugin/request-validation.t:360]: only one blank line is need at the end of file

File: apisix/plugins/request-validation.lua [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+    if conf.header_schema then
+        local ok, err = core.schema.check(conf.header_schema, headers)
+        if not ok then
+            core.log.error("req schema validation failed", err)
+            core.response.exit(400, err)
+        end
+    end
+
+    if conf.body_schema then
```
Comment (membphis) [inline apisix/plugins/request-validation.lua:60]: another style, I think this one is better:

```lua
if not conf.body_schema then
    return
end

ngx.req.read_body()
local body = ngx.req.get_body_data()
local req_body, err

if headers["content-type"] == "application/x-www-form-urlencoded" then
    req_body, err = ngx.decode_args(body)
else -- JSON as default
    req_body, err = core.json.decode(body)
end

if not req_body then
    ... ...
    return 
end

local ok, err = core.schema.check(conf.body_schema, req_body)
...
```

File: t/plugin/request-validation.t [hunk 13]
```diff
@@ -358,3 +358,33 @@ required field missing
...
+                ngx.say("failed")
+            end
+        }
+    }
+--- request
+GET /t
+--- response_body
+hello1 world
+--- no_error_log
+[error]
```
Comment (membphis) [inline t/plugin/request-validation.t:390]: need one blank line at the end of the file.

image

**Comment by membphis** [comment]
@sshniro you can rebase your branch with `master`, the CI will faster.

File: apisix/plugins/request-validation.lua (hunk 6 above)
Comment (sshniro) [reply apisix/plugins/request-validation.lua:45]: I do not understand this comment @membphis , ain't `coreschema.check` internally calls `create_validator` for validation?

Comment (membphis) [reply apisix/plugins/request-validation.lua:45]: here is the an valid example you provide, the user maybe provide a invalid schema:

```shell
"body_schema": {
    "type": "object",
    "required": ["required_payload"],
    "properties": {
            "emum_payload": {
            "type": "string",  # the user maybe specified a wrong type, eg: `str`
            "enum": ["enum_string_1", "enum_string_2"],
            "default": "enum_string_1"
        }
    }
}
```

here is detail:

image

File: doc/plugins/request-validation.md [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+    }
+}
+```
+
+
+## Examples:
+
+**Using ENUMS:**
+
+```shell
```
Comment (membphis) [inline doc/plugins/request-validation.md:110]: we should use `json` here

File: doc/plugins/request-validation.md [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+"body_schema": {
+    "type": "object",
+    "required": ["required_payload"],
+    "properties": {
+            "emum_payload": {
+            "type": "string",
+            enum: ["enum_string_1", "enum_string_2"]
```
Comment (membphis) [inline doc/plugins/request-validation.md:117]: this line should be wrong, it should be `"enum": `

File: doc/plugins/request-validation.md [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+            default = "enum_string_1"
+        }
+    }
+}
+```
+
+
+**JSON with multiple levels:**
+
+```shell
```
Comment (membphis) [inline doc/plugins/request-validation.md:127]: `json` ditto

**Comment by moonming** [comment]
we can merge first, then fix minor issues.

**Comment by membphis** [comment]
@sshniro merged, many thx

**Pull request merged**